src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.ReferenceCountedDisposable.cs

```diff
@@ -0,0 +1,225 @@
+            {
+                lock (referenceCount)
+                {
+                    if (referenceCount.Value == 0)
+                    {
+                        // The target is already disposed, and cannot be reused
+                        return null;
+                    }
+
+                    referenceCount.Value++;
+
+                    // Must return a new instance, in order for the Dispose operation on each individual instance to
+                    // be idempotent.
+                    return new ReferenceCountedDisposable<T>(target, referenceCount);
+                }
+            }
+
+            /// <summary>
+            /// Releases the current reference, causing the underlying object to be disposed if this was the last
+            /// reference.
+            /// </summary>
+            /// <remarks>
+            /// <para>After this instance is disposed, the <see cref="TryAddReference"/> method can no longer be used to
+            /// object a new reference to the target, even if other references to the target object are still in
+            /// use.</para>
+            /// </remarks>
+            public void Dispose()
+            {
+                T instanceToDispose = null;
+                lock (_boxedReferenceCount)
+                {
+                    if (_instance == null)
+                    {
+                        // Already disposed; allow multiple without error.
+                        return;
+                    }
+
+                    _boxedReferenceCount.Value--;
+                    if (_boxedReferenceCount.Value == 0)
+                    {
+                        instanceToDispose = _instance;
+                    }
+
+                    // Ensure multiple calls to Dispose for this instance are a NOP.
+                    _instance = null;
+                }
+
+                instanceToDispose?.Dispose();
+            }
+
+            /// <summary>
+            /// Represents a weak reference to a <see cref="ReferenceCountedDisposable{T}"/> which is capable of
+            /// obtaining a new counted reference up until the point when the object is no longer accessible.
+            /// </summary>
+            public struct WeakReference
+            {
+                /// <summary>
+                /// DO NOT DISPOSE OF THE TARGET.
+                /// </summary>
+                private readonly WeakReference<T> _instance;
+                private readonly StrongBox<int> _boxedReferenceCount;
+
+                public WeakReference(ReferenceCountedDisposable<T> reference)
+                    : this()
+                {
+                    var instance = reference._instance;
+                    var referenceCount = reference._boxedReferenceCount;
+                    if (instance == null || referenceCount == null)
+                    {
+                        // The specified reference is already not valid.
+                        return;
+                    }
```

Comment:
1. No, it's not possible. I can remove this check.
2. It would be equivalent to `new System.WeakReference(null)`. Costs an allocation but in the end `TryAddReference` will return null the same way it would any other time the target is disposed.

> Or will this code only get it once the ref count has already been incremented? If so, can we assert anything about the refcount here?

This is a weak reference; it never owns a [strong] reference. All values of `referenceCount.Value` are valid at this point, so the assertion would be trivially true.